feat: Extend GraphQL class to support multiple authentication methods…

timmy61109 opened the pull request:
… and more API options

This commit extends the constructor (`__init__` method) of the `GraphQL` class to handle a wider range of interactions with the GitLab API. The main changes include:

- **Added support for multiple authentication tokens:** In addition to the original `token`, it now supports `private_token`, `oauth_token`, and `job_token`, allowing users to configure based on different authentication requirements.
- **Added HTTP Basic Authentication:** Introduced `http_username` and `http_password` parameters, allowing the use of HTTP Basic Authentication to interact with the API.
- **Added API version control:** Introduced the `api_version` parameter, with a default value of `"4"`, allowing users to specify the GitLab API version to use.
- **Added pagination control parameters:** Added `per_page`, `pagination`, and `order_by` parameters to provide finer control over the API's pagination behavior and data sorting.
- **Added an option to keep the Base URL:** Added the `keep_base_url` parameter, allowing the configured base URL to be retained across multiple API calls.
- **Added `**kwargs: Any`:** Allows passing additional keyword arguments, providing greater flexibility.

These changes make the `GraphQL` class more powerful and flexible, capable of handling a broader range of GitLab API use cases.

<!-- Please make sure your commit messages follow Conventional Commits
(https://www.conventionalcommits.org) with a commit type (e.g. feat, fix, refactor, chore):

Bad:        Added support for release links
Good:     feat(api): add support for release links

Bad:        Update documentation for projects
Good:     docs(projects): update example for saving project attributes-->

## Changes

<!-- Remove this comment and describe your changes here. -->

### Documentation and testing

Please consider whether this PR needs documentation and tests. **This is not required**, but highly appreciated:

- [ ] Documentation in the matching [docs section](https://github.com/python-gitlab/python-gitlab/tree/main/docs)
- [ ] [Unit tests](https://github.com/python-gitlab/python-gitlab/tree/main/tests/unit) and/or [functional tests](https://github.com/python-gitlab/python-gitlab/tree/main/tests/functional)

<!--
Note: In some cases, basic functional tests may be easier to add, as they do not require adding mocked GitLab responses.
-->

JohnVillalovos commented on CHANGELOG.md:
```diff
@@ -1,5 +1,18 @@
+### Features
+
+- **group**: Add can use config file.
+  ([`bf7eba5`](https://github.com/python-gitlab/python-gitlab/pull/3177/commits/bf7eba5909669e19757ef1c5b590f613be00d5f4))
```
Semantic release will automatically create changelog entries. So should not modify this file.

JohnVillalovos commented on gitlab/_version.py:
```diff
@@ -3,4 +3,4 @@
 __license__ = "LGPL3"
 __title__ = "python-gitlab"
-__version__ = "5.6.0"
+__version__ = "5.7.1"
```
Semantic release will automatically update the version number. So should not modify this file.

timmy61109 replied:
Sorry, I made a change because Python 3 detection requires a new version number. I'll either remove this change before the official merge or rebase to your latest version and then use your version number.

timmy61109 commented:
> I'm not sure if some of or how much of this was made with the help of AI, but most of the changes do not apply to the GraphQL client. There is only a single way to authenticate against the API when using header auth, which we do (see https://docs.gitlab.com/api/graphql/#header-authentication). Likewise pagination and other things are not done the same way for GraphQL.
> 
> Could you maybe first elaborate on what you're trying to achieve and we can discuss this (possibly in an issue/discussion)?

裡面只有 `git commit` 是 AI 協助幫我翻譯完成，因為我的英文不是很好。為了可以快速可以用，因此我只是將 GitLab API Class 裡面功能參考與複製到 GraphQL API Class 裡面，並盡可能修改到可以使用。

我主要希望有以下目標：

1. 可以使用 GraphQL ，因為 GitLab 有些新功能 RSET API 目前尚未製作好，很多都是需要從 GraphQL 取得
2. 可以直接透過 `from_config` 取得資料，其實這個是我目前開這個合併請求的主要原因，我可以直接從 `gl.cfg` 匯入金鑰，這樣我不會疑問要將金鑰刪除
3. 由於現在 GraphQL 其實都是要自己寫指令，那這樣使用 Python 設計好取得資料的方式優勢就沒有，因此我希望未來可以有現成的 function 可以使用
4. 目前我製作的 function 是 timelogs ，我目前還在自我測試中，程式可能也沒有設計的很好，可以參考： https://gitlab.atcatw.org/atca/community-edition/workflow-automation-design/-/blob/9-csv-3/workflow_automation_design/expenses.py?ref_type=heads

---

Only the `git commit` part was translated with AI assistance because my English isn't very good. To make it usable quickly, I simply referenced and copied the functionality from the GitLab API Class into the GraphQL API Class and modified it as much as possible to make it usable.

My main goals are as follows:

1. To use GraphQL because some of GitLab's new features, such as the RSET API, are not yet available, and many require data to be retrieved from GraphQL.
2. To directly retrieve data via `from_config`. This is actually the main reason I opened this merge request. I can directly import the key from `gl.cfg`, so I won't have to worry about deleting the key.
3. Since GraphQL currently requires writing custom queries, the advantage of using Python to design data retrieval methods is lost. Therefore, I hope there will be pre-built functions available in the future.
4. The function I've created so far is `timelogs`. I'm still testing it myself, and the code might not be well-designed yet. You can check it out here: https://gitlab.atcatw.org/atca/community-edition/workflow-automation-design/-/blob/9-csv-3/workflow_automation_design/expenses.py?ref_type=heads

Translated with [DeepL.com](https://www.deepl.com/?utm_campaign=product&utm_source=web_translator&utm_medium=web&utm_content=copy_free_translation) (free version)

---

Only `git commit` was translated by AI because my English is not very good. In order to make it usable quickly, I just copied the function references in GitLab API Class to GraphQL API Class and modified it as much as possible to make it usable.

I mainly hope to achieve the following goals:

1. Use GraphQL, because some new features of GitLab, RSET API, are not yet ready, and many of them need to be obtained from GraphQL
2. Data can be obtained directly through `from_config`. In fact, this is the main reason why I opened this merge request. I can import the key directly from `gl.cfg`, so I don’t have to worry about deleting the key
3. Since GraphQL actually requires you to write your own instructions, there is no advantage in using Python to design a way to obtain data. Therefore, I hope that there will be ready-made functions available in the future
4. The function I have created is timelogs. I am still testing it myself, and the program may not be designed very well. You can refer to: https://gitlab.atcatw.org/atca/community-edition/workflow-automation-design/-/blob/9-csv-3/workflow_automation_design/expenses.py?ref_type=heads

Translated with Google Translate

timmy61109 commented:
You've discovered a great improvement! Adding `variable_values` to the `execute` function significantly enhances its flexibility and reusability. This allows for dynamic input without altering the core logic.

```python
gq.execute(  # type: ignore[attr-defined]
                query,
                variable_values=self.variable_values,
            )
```

Execution status:

```GraphQL
query get_timeslogs(
        $group_fullpath: ID! = "atca",
        $projects_first: Int = 1,
        $projects_after: String = "eyJpZCI6IjQyMCJ9",
        $timelogs_first: Int = 100,
        $timelogs_after: String = ""
        ) {
    group(fullPath: $group_fullpath) {
        projects(
                first: $projects_first,
                after: $projects_after,
                includeSubgroups: true
                ) {
            count
            nodes {
                timelogs(first: $timelogs_first, after: $timelogs_after) {
                    totalSpentTime
                    nodes {
                        id
                        mergeRequest {

                            projectId
                            id
                            name
                            iid
                            webPath
                            webUrl
                            state

                            timeEstimate
                            totalTimeSpent
                            humanTimeEstimate
                            humanTotalTimeSpent

                            labels {
                                edges {
                                    node {
                                        title
                                    }
                                }
                            }
                            author {
                                username
                            }
                            assignees(after: "", first: 100) {
                                nodes {
                                    username
                                }
                            }

                            createdAt
                            closedAt
                            updatedAt
                            mergedAt
                            preparedAt

                            approved
                            draft
                            mergeStatusEnum
                            mergeOngoing
                            mergeableDiscussionsState

                            reviewers {
                                edges {
                                    node {
                                        username
                                    }
                                }
                            }
                            approvedBy {
                                edges {
                                    node {
                                        username
                                    }
                                }
                            }
                            mergeUser {
                                username
                            }
                        }
                        spentAt
                        summary
                        timeSpent
                        user {
                            username
                        }
                      
                    }
                    pageInfo {
                        endCursor
                      	startCursor
                        hasNextPage
                      	hasPreviousPage
                    }
                }
            }
            pageInfo {
                endCursor
                startCursor
                hasNextPage
                hasPreviousPage
            }
        }
    }
}

```

Execution Result:

```
{
  "data": {
    "group": {
      "projects": {
        "count": null,
        "nodes": [
          {
            "timelogs": {
              "totalSpentTime": null,
              "nodes": [
                {
                  "id": null,
                  "mergeRequest": null,
                  "spentAt": null,
                  "summary": null,
                  "timeSpent": null,
                  "user": {
                    "username": null
                  }
                },
                {
                  "id": null,
                  "mergeRequest": null,
                  "spentAt": null,
                  "summary": null,
                  "timeSpent": null,
                  "user": {
                    "username": null
                  }
                },
                {
                  "id": null,
                  "mergeRequest": null,
                  "spentAt": null,
                  "summary": null,
                  "timeSpent": null,
                  "user": {
                    "username": null
                  }
                },
                {
                  "id": null,
                  "mergeRequest": null,
                  "spentAt": null,
                  "summary": null,
                  "timeSpent": null,
                  "user": {
                    "username": null
                  }
                }
              ],
              "pageInfo": {
                "endCursor": null,
                "startCursor": null,
                "hasNextPage": null,
                "hasPreviousPage": null
              }
            }
          }
        ],
        "pageInfo": {
          "endCursor": null,
          "startCursor": null,
          "hasNextPage": null,
          "hasPreviousPage": null
        }
      }
    }
  },
  "correlationId": null
}
```
